File: generated-code/example/ruby/model/one_of.rb
class OneOf
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == OneOf::OptionOne::TAG
            return OneOf::OptionOne.read_from(stream)
        end
        if tag == OneOf::OptionTwo::TAG
            return OneOf::OptionTwo.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    class OptionOne
        TAG = 0
    
        attr_accessor :vec_int
        attr_accessor :long_int
    
        def initialize(vec_int, long_int)
            @vec_int = vec_int
            @long_int = long_int
        end
    
        def self.read_from(stream)
            vec_int = []
            stream.read_int().times do |_|
                vec_int_element = stream.read_int()
                vec_int.push(vec_int_element)
            end
            long_int = stream.read_long()
            OptionOne.new(vec_int, long_int)
        end
    
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@vec_int.length())
            @vec_int.each do |vec_int_element|
                stream.write_int(vec_int_element)
            end
            stream.write_long(@long_int)
        end
    
        def to_s
            string_result = "OptionOne { "
            string_result += "vec_int: "
            string_result += "[ "
            vec_int_index = 0
            @vec_int.each do |vec_int_element|
                if vec_int_index != 0
                    string_result += ", "
                end
                string_result += vec_int_element.to_s
                vec_int_index += 1
            end
            string_result += " ]"
            string_result += ", "
            string_result += "long_int: "
            string_result += @long_int.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    class OptionTwo
        TAG = 1
    
        attr_accessor :value
    
        def initialize(value)
            @value = value
        end
    
        def self.read_from(stream)
            value = stream.read_int()
            OptionTwo.new(value)
        end
    
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@value)
        end
    
        def to_s
            string_result = "OptionTwo { "
            string_result += "value: "
            string_result += @value.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
end

File: generated-code/example/scala/src/main/scala/model/OneOf.scala
package model

import util.StreamUtil

sealed trait OneOf {
    def writeTo(stream: java.io.OutputStream)
}

object OneOf {
    case class OptionOne(vecInt: Seq[Int], longInt: Long) extends OneOf {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, OptionOne.TAG)
            StreamUtil.writeInt(stream, vecInt.length)
            vecInt.foreach { value =>
                StreamUtil.writeInt(stream, value)
            }
            StreamUtil.writeLong(stream, longInt)
        }
    
        override def toString(): String = {
            var stringBuilder = new StringBuilder("OptionOne { ")
            stringBuilder.append("vecInt: ")
            stringBuilder.append(vecInt)
            stringBuilder.append(", ")
            stringBuilder.append("longInt: ")
            stringBuilder.append(longInt)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object OptionOne {
        val TAG: Int = 0
    
        def readFrom(stream: java.io.InputStream): OptionOne = OptionOne(
            (0 until StreamUtil.readInt(stream)).map { _ =>
                StreamUtil.readInt(stream)
            },
            StreamUtil.readLong(stream)
        )
    }

    case class OptionTwo(value: Int) extends OneOf {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, OptionTwo.TAG)
            StreamUtil.writeInt(stream, value)
        }
    
        override def toString(): String = {
            var stringBuilder = new StringBuilder("OptionTwo { ")
            stringBuilder.append("value: ")
            stringBuilder.append(value)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object OptionTwo {
        val TAG: Int = 1
    
        def readFrom(stream: java.io.InputStream): OptionTwo = OptionTwo(
            StreamUtil.readInt(stream)
        )
    }

    def readFrom(stream: java.io.InputStream): OneOf = {
        StreamUtil.readInt(stream) match {
            case OptionOne.TAG => OptionOne.readFrom(stream)
            case OptionTwo.TAG => OptionTwo.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}